python/ray/data/dataset.py

```diff
@@ -854,6 +854,91 @@ def aggregate(self, *aggs: AggregateFn) -> U:
         ret = self.groupby(None).aggregate(*aggs).take(1)
         return ret[0] if len(ret) > 0 else None
 
+    def _check_and_normalize_agg_on(self,
+                                    on: AggregateOnT,
+                                    keys: Optional[List[str]] = None
+                                    ) -> AggregateOnT:
+        """Checks whether the provided aggregation `on` arg is valid for this
+        type of dataset, and normalizes the value based on the Dataset type and
+        any provided groupby keys.
+        """
+        if (on is not None
+                and (not isinstance(on, (str, Callable, list)) or
+                     (isinstance(on, list)
+                      and not (all(isinstance(on_, str) for on_ in on)
+                               or all(isinstance(on_, Callable)
+                                      for on_ in on))))):
+            raise TypeError(
+                f"`on` must be of type {AggregateOnT}, but got {type(on)}")
+
+        if isinstance(on, list) and len(on) == 0:
+            raise ValueError(
+                "When giving a list for `on`, it must be nonempty.")
+
+        # Determine if Dataset is an Arrow Dataset.
+        is_arrow = False
+        schema = None
+        try:
+            import pyarrow as pa
+        except ModuleNotFoundError:
+            is_arrow = False
+        else:
+            # We need schema to properly validate, so synchronously
+            # fetch it if necessary.
+            schema = self.schema(fetch_if_missing=True)
+            if schema is None:
+                return on
+            if isinstance(schema, pa.Schema):
+                is_arrow = True
+            else:
+                is_arrow = False
```

Comment:
Seems this is generally useful,  maybe extract into a helper function `def is_arrow_dataset()`?

Reply:
I originally tried to do that, but didn't want to fetch the schema multiple times, which would be required for both the `is_arrow_dataset()` and the remainder of this function. But since the schema fetch is cached, I think that multiple calls will be fine. I'll make this change!